kernel/src/transaction.rs

```diff
@@ -112,6 +138,107 @@ impl Transaction {
         self.commit_info = Some(commit_info.into());
         self
     }
+
+    // Generate the logical-to-physical transform expression which must be evaluated on every data
+    // chunk before writing. At the moment, this is a transaction-wide expression.
+    fn generate_logical_to_physical(&self) -> Expression {
+        // for now, we just pass through all the columns except partition columns.
+        // note this is _incorrect_ if table config deems we need partition columns.
+        Expression::struct_from(self.read_snapshot.schema().fields().filter_map(|f| {
+            if self
+                .read_snapshot
+                .metadata()
+                .partition_columns
+                .contains(f.name())
+            {
+                None
+            } else {
+                let col_name = ColumnName::new([f.name()]);
+                Some(col_name.into())
+            }
+        }))
+    }
+
+    /// Get the write context for this transaction. At the moment, this is constant for the whole
+    /// transaction.
+    // Note: after we introduce metadata updates (modify table schema, etc.), we need to make sure
+    // that engines cannot call this method after a metadata change, since the write context could
+    // have invalid metadata.
+    pub fn get_write_context(&self) -> WriteContext {
+        let target_dir = self.read_snapshot.table_root();
+        let snapshot_schema = self.read_snapshot.schema();
+        let logical_to_physical = self.generate_logical_to_physical();
+        WriteContext::new(
+            target_dir.clone(),
+            Arc::new(snapshot_schema.clone()),
+            logical_to_physical,
+        )
+    }
+
+    /// Add write metadata about files to include in the transaction. This API can be called
+    /// multiple times to add multiple batches.
+    ///
+    /// The expected schema for `write_metadata` is given by [`get_write_metadata_schema`].
+    pub fn add_write_metadata(&mut self, write_metadata: Box<dyn EngineData>) {
```

Comment:
I originally thought that most of our API used `Arc<dyn EngineData>` instead of box... but a quick search says we only have one Arc -- the transaction's `commit_info`. Maybe we should fix that other one?

Reply:
Oh interesting - I'll have that as another little follow-up: #464